src/RpcClient.ts

```diff
@@ -83,21 +83,21 @@ export abstract class RpcClient {
     }
 }
 
-export class PostMessageRpcClient extends RpcClient {
+class PostMessageRpcClient extends RpcClient {
     private _target: Window | null;
     private readonly _receiveListener: (message: MessageEvent) => any;
-    private _connected: boolean;
+    private _connectionState: PostMessageRpcClient.ConnectionState;
 
     constructor(targetWindow: Window, allowedOrigin: string) {
         super(allowedOrigin);
         this._target = targetWindow;
-        this._connected = false;
+        this._connectionState = PostMessageRpcClient.ConnectionState.DISCONNECTED;
 
         this._receiveListener = this._receive.bind(this);
     }
 
     public async init() {
-        if (this._connected) {
+        if (this._connectionState === PostMessageRpcClient.ConnectionState.CONNECTED) {
```

Comment:
Now that you have three states, instead of two, are you sure you have not forgotten any checks for the third state? Like here, do you not need to check if it's already `CONNECTING`?

Reply:
Yes, I think I considered that everywhere.
Here I'm not checking for `CONNECTING` on purpose because if a caller calls `await client.init()` he expects that initialization successfully finished after the await which would not be the case when returning here in `CONNECTING` state.
The way the method is implemented, it's safe to call `init` multiple times in parallel, including during `CONNECTING` state.